[Dynamo][Better Engineering] Add typing annotations to guard and source

Lucaskabela opened the pull request:
As part of better engineering week, we would like to improve out type support to improve dev experience in dynamo

This PR adds strict typing support to a critical set of files for dynamo, `source.py` and the base `_guards.py`

Running 
```
mypy torch/_dynamo/source.py torch/_guards.py --linecount-report /tmp/coverage_log
```

| -------- | Lines Unannotated | Lines Total | % lines covered | Funcs Unannotated | Funcs Total | % funcs covered |
| -------- | ------- | -------- | ------- | ------- | ------- | ------- |
| Main  |  1227 | 2208 | 55.57% | 207 | 362 | 57.18% |
| This PR | 2217 | 2217 | 100.00% | 362 | 362 | 100.00% |
| Delta    | +990 | +9 | +44.43% | +155 | 0 | +42.82% |



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Lucaskabela commented:
@pytorchbot label "topic: not user facing"

Lucaskabela commented:
Note for reviewers - `torch/_dynamo/source.py` and `torch/_guards.py` contain primary changes; changes to other files are mostly to handle/check for None from optional returns to make type checker happy and avoid errors

anijain2305 reviewed (approved):
Thanks for working on this!

Lucaskabela commented:
@pytorchmergebot merge

Lucaskabela commented:
Error is unrelated, from https://github.com/pytorch/pytorch/pull/158750

Lucaskabela commented:
@pytorchmergebot merge

Lucaskabela commented:
@pytorchbot revert -m "Suggested to fix failing internal signals on D78911890" -c nosignal

yangw-dev commented:
@pytorchbot revert -m "Suggested to fix failing internal signals on D78911890" -c nosignal

Lucaskabela committed:
[Dynamo][Better Engineering] Add typing annotations to guard and source (#158397)

Summary:
X-link: https://github.com/pytorch/executorch/pull/12986

As part of better engineering week, we would like to improve out type support to improve dev experience in dynamo

This PR adds strict typing support to a critical set of files for dynamo, `source.py` and the base `_guards.py`

Running
```
mypy torch/_dynamo/source.py torch/_guards.py --linecount-report /tmp/coverage_log
```

| -------- | Lines Unannotated | Lines Total | % lines covered | Funcs Unannotated | Funcs Total | % funcs covered |
| -------- | ------- | -------- | ------- | ------- | ------- | ------- |
| Main  |  1227 | 2208 | 55.57% | 207 | 362 | 57.18% |
| This PR | 2217 | 2217 | 100.00% | 362 | 362 | 100.00% |
| Delta    | +990 | +9 | +44.43% | +155 | 0 | +42.82% |



cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 jerryzh168 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben


Test Plan:
Imported from GitHub, without a `Test Plan:` line.

Rollback Plan:

Reviewed By: yangw-dev

Differential Revision: D79199389

Pulled By: Lucaskabela
